packages/store/addon/-private/system/model/internal-model.ts

```diff
@@ -435,24 +418,41 @@ export default class InternalModel {
     });
 
     this._record = null;
-    this.isReloading = false;
     this.error = null;
+    this._previousState = this.currentState;
     this.currentState = RootState.empty;
     if (REMOVE_RECORD_ARRAY_MANAGER_LEGACY_COMPAT) {
       this.store.recordArrayManager.recordDidChange(this.identifier);
     }
   }
 
   deleteRecord() {
-    if (RECORD_DATA_STATE) {
-      if (this._recordData.setIsDeleted) {
-        this._recordData.setIsDeleted(true);
-      }
-    }
-    this.send('deleteRecord');
+    run(() => {
```

Comment:
Why the double runloop?

Reply:
outer one is Ember's which ensure the destroy queue will be completed, inner one is ours which ensures the relationship sweep will be completed. Both are necessary to ensure that deleteRecord completes this state updates synchronously.

Reply:
But delete record doesn't cause any destruction? Wouldn't this be fully encompassed by the call to `unloadRecord`?

Reply:
@igorT it does when a new record is deleted.